core/src/main/scala/com/mesosphere/usi/core/matching/RangeResource.scala

```diff
@@ -0,0 +1,251 @@
+
+    if (numberOfOfferedValues == 0) {
+      return Iterator.empty
+    }
+
+    def findStartValue(shuffled: IndexedSeq[MesosRange], startValueIdx: Int): (Int, Int) = {
+      var startValueIdxOfCurrentRange = 0
+      val rangeIdx = shuffled.indexWhere {
+        case range: MesosRange if startValueIdxOfCurrentRange + range.size > startValueIdx =>
+          true
+        case range: MesosRange =>
+          startValueIdxOfCurrentRange += range.size
+          false
+      }
+
+      (rangeIdx, startValueIdx - startValueIdxOfCurrentRange)
+    }
+
+    val shuffled = rand.shuffle(ranges).toIndexedSeq
+    val startValueIdx = rand.nextInt(numberOfOfferedValues)
+    val (rangeIdx, valueInRangeIdx) = findStartValue(shuffled, startValueIdx)
+    val startRangeOrig = shuffled(rangeIdx)
+
+    val startRange = startRangeOrig.drop(valueInRangeIdx)
+
+    // These are created on demand if necessary
+    def afterStartRange: Iterator[Int] =
+      shuffled.slice(rangeIdx + 1, shuffled.length).iterator.flatMap(_.iterator)
+    def beforeStartRange: Iterator[Int] =
+      shuffled.slice(0, rangeIdx).iterator.flatMap(_.iterator)
+    def endRange: Iterator[Int] = startRangeOrig.take(valueInRangeIdx)
+
+    startRange ++ afterStartRange ++ beforeStartRange ++ endRange
+  }
+}
+
+case class MesosRange(minValue: Int, maxValue: Int) {
+  private[this] def range: Range.Inclusive = Range.inclusive(minValue, maxValue)
+  def size: Int = range.size
+
+  def iterator: Iterator[Int] = range.iterator.map(v => v)
+  def drop(n: Int): Iterator[Int] = range.drop(n).iterator.map(v => v)
+  def take(n: Int): Iterator[Int] = range.take(n).iterator.map(v => v)
+
+  /*
+   * Attention! range exports _two_ contains methods, a generic inefficient one and an efficient one
+   * that only gets used with Int (and not java.lang.Integer and similar)
+   */
+  def contains(v: Int): Boolean = range.contains(v)
+}
+
+sealed trait ValueMatchResult
+case object ValueNotAvailable extends ValueMatchResult
+case class ValueMatched(port: Int) extends ValueMatchResult
+
+object RangeResource {
+  val RandomValue: Int = 0
+
+  def ports(requestedPorts: Seq[Int], random: Random = Random): RangeResource = {
+    new RangeResource(requestedPorts, ResourceType.PORTS, random)
+  }
+
+  /**
+    * Return RangesResources covering all given values with the given roles.
+    *
+    * Creates as few RangesResources as possible while
+    * preserving the order of the values.
+    */
+  def createMesosResource(
+      resourceFromOffer: Protos.Resource,
+      requestedValues: Seq[Int],
+      resourceType: ResourceType): Seq[Protos.Resource] = {
+    /*
+     * Create as few ranges as possible from the given values while preserving the order of the values.
+     *
+     * It does not check if the given values have different roles.
+     */
+    def createRanges(ranges: Seq[Int]): Seq[Range] = {
+      val builder = Seq.newBuilder[Range]
+
+      @tailrec
```

Comment:
The builder pattern and tail recursion don't mix well IMHO. This should be a simple iteration or fold.

```scala
val builder = Vector.newBuilder[Range]

if (ranges.nonEmpty) {
  var currentRange = ranges.head
  ranges.tail.foreach { v =>
    if (overlaps(currentRange, v)) {
     currentRange = Range(currentRange.first, v)
    } else {
      builder += currentRange
      currentRange = Range(v, v)
    }
  }
  builder += currentRange // append last range.
}
```

```scala
ranges.foldLeft(Vector.empty[Range]) { (acc, next) =>
  // Merge next range and last if they overlap
  if (acc.headOption.exists(curr => overlaps(curr, next))) {
    Range(acc.head.first, next) +: acc.tail
  } else {
    Range(next, next) +: acc
  }
}
```

Anyhow, are the ranges always sorted?

Reply:
"while preserving the order of the values."

Reply:
and agreed, I chose the iterative way an re-implemented it. Thanks!